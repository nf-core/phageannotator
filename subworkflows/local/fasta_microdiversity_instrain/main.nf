//
// Assess virus microdiversity with instrain
//

include { INSTRAIN_PROFILE          } from '../../../modules/nf-core/instrain/profile/main'
include { INSTRAIN_COMPARE          } from '../../../modules/nf-core/instrain/compare/main'


workflow FASTA_MICRODIVERSITY_INSTRAIN {
    take:
    bam             // [ [ meta ], bam ]        , BAM files from reads aligned to FASTA file (mandatory)
    genome_fasta    // [ [ meta ], fasta ]      , FASTA file used in read alignment (mandatory)
    proteins_fna    // [ [ meta ], fna ]        , FASTA file for protein-coding genes (optional)
    instrain_stb    // [ [ meta ], stb.tsv ]    , TSV file with two columns for associationg scaffolds to bins (optional)

    main:
    ch_versions = Channel.empty()

    // remove meta information for instrain profile inputs
    ch_genome_fasta_nometa = genome_fasta.map { it[1] }.first()
    if ( proteins_fna ){
        ch_proteins_fna_nometa = proteins_fna.map { it[1] }.first()
    } else {
        ch_proteins_fna_nometa = []
    }
    if ( instrain_stb ){
        ch_stb_file_tsv_nometa = instrain_stb.map { it[1] }.first()
    } else {
        ch_stb_file_tsv_nometa = []
    }

    //
    // MODULE: Profile microdiveristy within each sample
    //
    ch_instrain_profiles = INSTRAIN_PROFILE ( bam, ch_genome_fasta_nometa, ch_proteins_fna_nometa, ch_stb_file_tsv_nometa ).profile
    ch_instrain_gene_tsv = INSTRAIN_PROFILE.out.gene_info
    ch_versions = ch_versions.mix(INSTRAIN_PROFILE.out.versions)

    // combine bams and profiles across samples
    ch_bam_profiles_combined = bam.join(ch_instrain_profiles)
    ch_instrain_compare_input = ch_bam_profiles_combined.map { [ [ id:'all_samples' ], it[1], it[2] ] }.groupTuple( sort: 'deep' )

    //
    // MODULE: Compare microdiversity across samples (within groups OR across all samples)
    //
    ch_instrain_compare = INSTRAIN_COMPARE ( ch_instrain_compare_input, ch_stb_file_tsv_nometa ).compare
    ch_instrain_comparisons_tsv = INSTRAIN_COMPARE.out.comparisons_table
    ch_instrain_pooled_snvs_tsv = INSTRAIN_COMPARE.out.pooled_snv

    ch_versions = ch_versions.mix(INSTRAIN_COMPARE.out.versions)

    emit:
    gene_info_tsv   = ch_instrain_gene_tsv  // [ [ meta ], gene_info.tsv ]
    versions        = ch_versions           // [ versions.yml ]
}
