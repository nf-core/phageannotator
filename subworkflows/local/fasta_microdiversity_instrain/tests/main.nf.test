nextflow_workflow {

    name "Test Subworkflow: FASTA_MICRODIVERSITY_INSTRAIN"
    script "../main.nf"
    workflow "FASTA_MICRODIVERSITY_INSTRAIN"

    // Dependencies
    tag "INSTRAIN_PROFILE"
    tag "INSTRAIN_COMPARE"

    test("bam & fasta & stb_file") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                    ],
                    [
                        [ id:'test2' ],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.sorted.bam', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of([
                    [ id: 'reference_fasta' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                    ]
                )
                input[2] = null
                input[3] = Channel.of([
                    [ id: 'instrain_stb' ],
                    file(params.pipelines_testdata_base_path + 'modules/nfcore/instrain/profile/instrain_stb.stb', checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
