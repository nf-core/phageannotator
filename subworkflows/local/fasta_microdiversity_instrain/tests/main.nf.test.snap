{
    "bam & fasta & stb_file": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,816a6bc0dc4c7fe8dee2a3cd012b38e2"
                ],
                "gene_info_tsv": [
                    
                ],
                "versions": [
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,4f3be9f56f69dcce1903598ad2fc9635",
                    "versions.yml:md5,816a6bc0dc4c7fe8dee2a3cd012b38e2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-08T20:15:50.456499453"
    }
}